### PR TITLE
188: support Firefox and Microsoft Edge

### PR DESCRIPTION
**Pull request by mdiwakar**
188: support Firefox and Microsoft Edge capabilities on testcontainers execution

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.*

Please provide enough information so that others can review your pull request:

Explain the **details** for making this change. What existing problem does the pull request solve?

**Test plan (required)**

Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.

**Code formatting**

**Closing issues**

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

**Comment by mdiwakar** [comment]
@eliasnogueira , I have refactored the code to support Firefox  Microsoft Edge capabilities on test containers, please review

**Comment by mdiwakar** [comment]
closes 188

File: .sdlc/Jenkinsfile [hunk 1]
```diff
@@ -2,7 +2,7 @@ node {
     def mvnHome
 
     stage('Preparation') {
-        git 'https://github.com/eliasnogueira/selenium-java-lean-test-achitecture.git'
+        git 'https://github.com/mdiwakar/selenium-java-lean-test-achitecture.git'
```
Comment (eliasnogueira) [inline .sdlc/Jenkinsfile:5]: Could you revert the change to the original repo?

File: pom.xml [hunk 2]
```diff
@@ -9,8 +9,8 @@
     <version>3.3.1</version>
 
     <scm>
-        <connection>scm:git@github.com:eliasnogueira/selenium-java-lean-test-architecture.git</connection>
-        <developerConnection>scm:git@github.com:eliasnogueira/selenium-java-lean-test-architecture.git</developerConnection>
+        <connection>scm:git@github.com:mdiwakar/selenium-java-lean-test-architecture.git</connection>
```
Comment (eliasnogueira) [inline pom.xml:12]: Could you revert the change to the original repo?

File: src/main/java/com/eliasnogueira/driver/TargetFactory.java [hunk 3]
```diff
@@ -27,16 +27,19 @@
...
+import org.openqa.selenium.firefox.FirefoxOptions;
 import org.openqa.selenium.remote.RemoteWebDriver;
 import org.testcontainers.containers.BrowserWebDriverContainer;
 
 import java.net.URL;
 
 import static com.eliasnogueira.config.ConfigurationManager.configuration;
-import static com.eliasnogueira.driver.BrowserFactory.CHROME;
-import static com.eliasnogueira.driver.BrowserFactory.valueOf;
+import static com.eliasnogueira.driver.BrowserFactory.*;
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/TargetFactory.java:42]: Could you remove the start import and add the specialized ones?

File: src/test/resources/general.properties [hunk 4]
```diff
@@ -1,8 +1,8 @@
 # target execution: local, selenium-grid or testcontainers
-target = local
+target = testcontainers
```
Comment (eliasnogueira) [inline src/test/resources/general.properties:2]: Could you revert the change to the original value?

File: src/test/resources/general.properties [hunk 5]
```diff
@@ -1,8 +1,8 @@
...
 
 # browser to use for local and testcontainers execution
-browser = chrome
+browser = edge
```
Comment (eliasnogueira) [inline src/test/resources/general.properties:5]: Could you revert the change to the original value?

File: src/main/java/com/eliasnogueira/driver/TargetFactory.java [hunk 6]
```diff
@@ -74,16 +77,23 @@ private RemoteWebDriver createRemoteInstance(MutableCapabilities capability) {
 
     private RemoteWebDriver createTestContainersInstance(MutableCapabilities capabilities) {
         String browser = capabilities.getBrowserName();
+        BrowserWebDriverContainer<?> driverContainer = null ;
+        if (browser.equalsIgnoreCase( CHROME.toString())) {
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/TargetFactory.java:81]: I believe this could be part of the `BrowserFactory` class as you now have added the ability to use testcontainers for different browsers, so we would have a single source of truth.

You can follow the same approach applied to the local or remote driver.

**Review by eliasnogueira — CHANGES_REQUESTED** [review]

File: src/main/java/com/eliasnogueira/driver/TargetFactory.java (hunk 3 above)
Comment (mdiwakar) [reply src/main/java/com/eliasnogueira/driver/TargetFactory.java]: done

File: pom.xml (hunk 2 above)
Comment (mdiwakar) [reply pom.xml]: done

File: .sdlc/Jenkinsfile (hunk 1 above)
Comment (mdiwakar) [reply .sdlc/Jenkinsfile]: done

File: src/test/resources/general.properties (hunk 5 above)
Comment (mdiwakar) [reply src/test/resources/general.properties]: done

**Comment by mdiwakar** [comment]
@eliasnogueira , made changes according to review, please check and merge the PR

**Review by eliasnogueira — CHANGES_REQUESTED** [review]
> @eliasnogueira , made changes according to review, please check and merge the PR

Hi @mdiwakar ,
I can see that the tests targeting Chrome engine (Google Chrome and Edge) are failing.

This is the error:
```
SessionNotCreated Could not start a new session. Response code 500. Message: Could not start a new session. Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: unknown error: Chrome failed to start: crashed.
```

Could you check it?

**Comment by mdiwakar** [comment]
@eliasnogueira , tests pass on my laptop for both local as well as test containers  'mvn test', can you please send me the command you used to run?

**Comment by eliasnogueira** [comment]
> @eliasnogueira , tests pass on my laptop for both local as well as test containers 'mvn test', can you please send me the command you used to run?

Of course. I'm using this command, changing the `-Dbrowser` to the different ones:
```
mvn clean test -D -Pweb-execution -Dtarget=testcontainers -Dbrowser=chrome
```

**Comment by mdiwakar** [comment]
PS C:\git\selenium-java-lean-test-architecture> mvn clean test -Pweb-execution -Dtarget=testcontainers -Dbrowser=chrome                       
[INFO] Scanning for projects...
[INFO] 
[INFO] -------< com.eliasnogueira:selenium-java-lean-test-architecture >-------
[INFO] Building selenium-java-lean-test-architecture 3.3.1
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ selenium-java-lean-test-architecture ---
[INFO] Deleting C:\git\selenium-java-lean-test-architecture\target
[INFO] 
[INFO] --- resources:3.3.0:resources (default-resources) @ selenium-java-lean-test-architecture ---
[INFO] Copying 1 resource
[INFO]
[INFO] --- compiler:3.10.1:compile (default-compile) @ selenium-java-lean-test-architecture ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 17 source files to C:\git\selenium-java-lean-test-architecture\target\classes
[INFO] 
[INFO] --- resources:3.3.0:testResources (default-testResources) @ selenium-java-lean-test-architecture ---
[INFO] Copying 5 resources
[INFO]
[INFO] --- compiler:3.10.1:testCompile (default-testCompile) @ selenium-java-lean-test-architecture ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to C:\git\selenium-java-lean-test-architecture\target\test-classes
[INFO] 
[INFO] --- surefire:3.0.0-M7:test (default-test) @ selenium-java-lean-test-architecture ---
[INFO] Using auto detected provider org.apache.maven.surefire.testng.TestNGProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running TestSuite
Allure environment data saved.
[INFO ] 2023-07-03 07:07:59.256 [TestNG-tests-1] ImageNameSubstitutor - Image name substitution will be performed by: DefaultImageNameSubstitutor (composite of 'ConfigurationFileImageNameSubstitutor' and 'PrefixingImageNameSubstitutor')
[INFO ] 2023-07-03 07:07:59.845 [TestNG-tests-1] DockerClientProviderStrategy - Loaded org.testcontainers.dockerclient.NpipeSocketClientProviderStrategy from ~/.testcontainers.properties, will try it first
[INFO ] 2023-07-03 07:08:00.592 [TestNG-tests-1] DockerClientProviderStrategy - Found Docker environment with local Npipe socket (npipe:////./pipe/docker_engine)
[INFO ] 2023-07-03 07:08:00.592 [TestNG-tests-1] DockerClientFactory - Docker host IP address is localhost
[INFO ] 2023-07-03 07:08:00.622 [TestNG-tests-1] DockerClientFactory - Connected to docker:
  Server Version: 23.0.5
  API Version: 1.42
  Operating System: Docker Desktop
  Total Memory: 7805 MB
[INFO ] 2023-07-03 07:08:00.684 [TestNG-tests-1] 1 - Creating container for image: testcontainers/ryuk:0.5.1
[INFO ] 2023-07-03 07:08:00.955 [TestNG-tests-1] RegistryAuthLocator - Credential helper/store (docker-credential-desktop) does not have credentials for https://index.docker.io/v1/
[INFO ] 2023-07-03 07:08:01.184 [TestNG-tests-1] 1 - Container testcontainers/ryuk:0.5.1 is starting: 7c8dba1964e6c1db6750ba7133f47f4da33e813a12687ec753f2eeb6aadd2290
[INFO ] 2023-07-03 07:08:02.788 [TestNG-tests-1] 1 - Container testcontainers/ryuk:0.5.1 started in PT2.1453262S
[INFO ] 2023-07-03 07:08:02.835 [TestNG-tests-1] RyukResourceReaper - Ryuk started - will monitor and terminate Testcontainers containers on JVM exit
[INFO ] 2023-07-03 07:08:02.835 [TestNG-tests-1] DockerClientFactory - Checking the system...
[INFO ] 2023-07-03 07:08:02.835 [TestNG-tests-1] DockerClientFactory - ?? Docker server version should be at least 1.6.0
[INFO ] 2023-07-03 07:08:02.835 [TestNG-tests-1] SeleniumUtils - Selenium API version 4.9.0 detected on classpath
[INFO ] 2023-07-03 07:08:03.023 [TestNG-tests-1] 0 - Creating container for image: selenium/standalone-chrome:4.9.0
[INFO ] 2023-07-03 07:08:03.354 [TestNG-tests-1] 0 - Container selenium/standalone-chrome:4.9.0 is starting: 22a7130f94919d545b58793fa787cc263e6134d0bf72b04b8c0496023594762b
[INFO ] 2023-07-03 07:08:08.871 [TestNG-tests-1] 0 - Container selenium/standalone-chrome:4.9.0 started in PT5.8486225S
[INFO ] 2023-07-03 07:08:08.873 [TestNG-tests-1] 0 - Creating container for image: testcontainers/vnc-recorder:1.3.0
[INFO ] 2023-07-03 07:08:08.950 [TestNG-tests-1] 0 - Container testcontainers/vnc-recorder:1.3.0 is starting: da704cb0662a48eff7f5d3bb7b4a552362ef77d2fd322ae29a80a99a40c5b2af
[INFO ] 2023-07-03 07:08:10.865 [TestNG-tests-1] 0 - Container testcontainers/vnc-recorder:1.3.0 started in PT1.9918084S
Jul 03, 2023 7:08:10 AM org.openqa.selenium.remote.tracing.opentelemetry.OpenTelemetryTracer createTracer
INFO: Using OpenTelemetry for tracing
[INFO ] 2023-07-03 07:08:15.150 [TestNG-tests-1] BookingDataFactory - Booking[email=christel.metz@yahoo.com, country=Belgium, password=sx2o88tui767n5ga, dailyBudget=
$499 - $999, newsletter=false, roomType=FAMILY, roomDescription=Eius voluptatem id molestiae cum odit. Qui cupiditate eos. Praesentium reiciendis ut sunt dolore error animi.]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 20.929 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  27.061 s
[INFO] Finished at: 2023-07-03T07:08:17-07:00
[INFO] ------------------------------------------------------------------------
[WARNING]
[WARNING] Plugin validation issues were detected in 3 plugin(s)
[WARNING]
[WARNING]  * org.apache.maven.plugins:maven-compiler-plugin:3.10.1
[WARNING]  * org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M7
[WARNING]  * org.apache.maven.plugins:maven-resources-plugin:3.3.0
[WARNING]
[WARNING] For more or less details, use 'maven.plugin.validation' property with one of the values (case insensitive): [BRIEF, DEFAULT, VERBOSE]
[WARNING]
PS C:\git\selenium-java-lean-test-architecture> ``

**Comment by mdiwakar** [comment]
`PS C:\git\selenium-java-lean-test-architecture> mvn clean test -Pweb-execution -Dtarget=testcontainers -Dbrowser=edge  
[INFO] Scanning for projects...
[INFO] 
[INFO] -------< com.eliasnogueira:selenium-java-lean-test-architecture >-------
[INFO] Building selenium-java-lean-test-architecture 3.3.1
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ selenium-java-lean-test-architecture ---
[INFO] Deleting C:\git\selenium-java-lean-test-architecture\target
[INFO] 
[INFO] --- resources:3.3.0:resources (default-resources) @ selenium-java-lean-test-architecture ---
[INFO] Copying 1 resource
[INFO]
[INFO] --- compiler:3.10.1:compile (default-compile) @ selenium-java-lean-test-architecture ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 17 source files to C:\git\selenium-java-lean-test-architecture\target\classes
[INFO] 
[INFO] --- resources:3.3.0:testResources (default-testResources) @ selenium-java-lean-test-architecture ---
[INFO] Copying 5 resources
[INFO] 
[INFO] --- compiler:3.10.1:testCompile (default-testCompile) @ selenium-java-lean-test-architecture ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to C:\git\selenium-java-lean-test-architecture\target\test-classes
[INFO] 
[INFO] --- surefire:3.0.0-M7:test (default-test) @ selenium-java-lean-test-architecture ---
[INFO] Using auto detected provider org.apache.maven.surefire.testng.TestNGProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running TestSuite
Allure environment data saved.
[INFO ] 2023-07-03 07:14:18.802 [TestNG-tests-1] ImageNameSubstitutor - Image name substitution will be performed by: DefaultImageNameSubstitutor (composite of 'ConfigurationFileImageNameSubstitutor' and 'PrefixingImageNameSubstitutor')
[INFO ] 2023-07-03 07:14:19.353 [TestNG-tests-1] DockerClientProviderStrategy - Loaded org.testcontainers.dockerclient.NpipeSocketClientProviderStrategy from ~/.testcontainers.properties, will try it first
[INFO ] 2023-07-03 07:14:20.060 [TestNG-tests-1] DockerClientProviderStrategy - Found Docker environment with local Npipe socket (npipe:////./pipe/docker_engine)
[INFO ] 2023-07-03 07:14:20.062 [TestNG-tests-1] DockerClientFactory - Docker host IP address is localhost
[INFO ] 2023-07-03 07:14:20.087 [TestNG-tests-1] DockerClientFactory - Connected to docker:
  Server Version: 23.0.5
  API Version: 1.42
  Operating System: Docker Desktop
  Total Memory: 7805 MB
[INFO ] 2023-07-03 07:14:20.137 [TestNG-tests-1] 1 - Creating container for image: testcontainers/ryuk:0.5.1
[INFO ] 2023-07-03 07:14:20.412 [TestNG-tests-1] RegistryAuthLocator - Credential helper/store (docker-credential-desktop) does not have credentials for https://index.docker.io/v1/
[INFO ] 2023-07-03 07:14:20.685 [TestNG-tests-1] 1 - Container testcontainers/ryuk:0.5.1 is starting: b508a4f325de53d63afea760a8107934e17c3207c3bdcf7f27858979b14864d3
[INFO ] 2023-07-03 07:14:22.500 [TestNG-tests-1] 1 - Container testcontainers/ryuk:0.5.1 started in PT2.3950698S
[INFO ] 2023-07-03 07:14:22.554 [TestNG-tests-1] RyukResourceReaper - Ryuk started - will monitor and terminate Testcontainers containers on JVM exit
[INFO ] 2023-07-03 07:14:22.555 [TestNG-tests-1] DockerClientFactory - Checking the system...
[INFO ] 2023-07-03 07:14:22.556 [TestNG-tests-1] DockerClientFactory - ?? Docker server version should be at least 1.6.0
[INFO ] 2023-07-03 07:14:22.559 [TestNG-tests-1] SeleniumUtils - Selenium API version 4.9.0 detected on classpath
[INFO ] 2023-07-03 07:14:22.785 [TestNG-tests-1] 0 - Creating container for image: selenium/standalone-edge:4.9.0
[INFO ] 2023-07-03 07:14:22.975 [TestNG-tests-1] 0 - Container selenium/standalone-edge:4.9.0 is starting: 84a4fe72d9b8a81cbedf73713ae3f8e00bd3595aba3082f06c9addd224ffe996
[INFO ] 2023-07-03 07:14:28.432 [TestNG-tests-1] 0 - Container selenium/standalone-edge:4.9.0 started in PT5.6482231S
[INFO ] 2023-07-03 07:14:28.433 [TestNG-tests-1] 0 - Creating container for image: testcontainers/vnc-recorder:1.3.0
[INFO ] 2023-07-03 07:14:28.516 [TestNG-tests-1] 0 - Container testcontainers/vnc-recorder:1.3.0 is starting: 5395f71482d0a743f773db15877ec4b561bcec615860d297fc58125ebde016ee
[INFO ] 2023-07-03 07:14:30.680 [TestNG-tests-1] 0 - Container testcontainers/vnc-recorder:1.3.0 started in PT2.2466151S
Jul 03, 2023 7:14:30 AM org.openqa.selenium.remote.tracing.opentelemetry.OpenTelemetryTracer createTracer
INFO: Using OpenTelemetry for tracing
[INFO ] 2023-07-03 07:14:35.012 [TestNG-tests-1] BookingDataFactory - Booking[email=jamar.schroeder@gmail.com, country=Netherlands, password=ak0o6os6, dailyBudget=$1
00, newsletter=false, roomType=BUSINESS, roomDescription=Consequatur asperiores inventore. Ipsa quas optio ut. Quia earum laudantium rerum eum. Delectus architecto cupiditate quia aut. Enim nobis tempore nemo sequi sunt et laudantium.]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 21.672 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------`

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java [hunk 7]
```diff
@@ -55,7 +59,15 @@ public WebDriver createLocalDriver() {
 
             return new ChromeDriver(getOptions());
         }
+        @Override
+        public WebDriver createTestContainerDriver() {
+            BrowserWebDriverContainer<?> driverContainer = new BrowserWebDriverContainer<>()
+                    .withCapabilities(new ChromeOptions());
+                driverContainer.start();
+                return new RemoteWebDriver(driverContainer.getSeleniumAddress(), new ChromeOptions());
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/BrowserFactory.java:68]: ## Fix

Can you remove the empty lines?

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java [hunk 8]
```diff
@@ -75,8 +87,16 @@ public WebDriver createLocalDriver() {
...
             return new FirefoxDriver(getOptions());
         }
-
         @Override
+        public WebDriver createTestContainerDriver() {
+            BrowserWebDriverContainer<?> driverContainer = new BrowserWebDriverContainer<>()
+                    .withCapabilities(new FirefoxOptions());
+                driverContainer.start();
+                return new RemoteWebDriver(driverContainer.getSeleniumAddress(), new FirefoxOptions());
+
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/BrowserFactory.java:96]: ## Fix

Can you remove the empty lines?

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java [hunk 9]
```diff
@@ -93,6 +113,14 @@ public WebDriver createLocalDriver() {
             return new EdgeDriver(getOptions());
         }
 
+        public WebDriver createTestContainerDriver() {
+            BrowserWebDriverContainer<?> driverContainer = new BrowserWebDriverContainer<>()
+                    .withCapabilities(new EdgeOptions());
+                driverContainer.start();
+                return new RemoteWebDriver(driverContainer.getSeleniumAddress(), new EdgeOptions());
+
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/BrowserFactory.java:121]: ## Fix

Can you remove the empty line?

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java [hunk 10]
```diff
@@ -109,6 +137,11 @@ public WebDriver createLocalDriver() {
 
             return new SafariDriver(getOptions());
         }
+        public WebDriver createTestContainerDriver() {
+            throw new IllegalArgumentException(
+                    "Browser Safari not supported on TestContainers yet");
+
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/BrowserFactory.java:143]: ## Fix

Can you remove the empty lines?

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java [hunk 11]
```diff
@@ -133,4 +166,8 @@ public SafariOptions getOptions() {
      * @return a new AbstractDriverOptions instance based on the browser set
      */
     public abstract AbstractDriverOptions<?> getOptions();
+
+
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/BrowserFactory.java:170]: ## Fix

Can you remove the empty line?

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java [hunk 12]
```diff
@@ -55,7 +59,15 @@ public WebDriver createLocalDriver() {
 
             return new ChromeDriver(getOptions());
         }
+        @Override
+        public WebDriver createTestContainerDriver() {
```
Comment (eliasnogueira) [inline src/main/java/com/eliasnogueira/driver/BrowserFactory.java:63]: ## Fix

Can you apply the code format in this class to follow the correct code indentation?

**Review by eliasnogueira — CHANGES_REQUESTED** [review]
Hey!
I've tested again and found out that the tests aren't running because of the M2 chip.

Would you mind adding the following note in the readme file for the testcontainers section?

```
If you are using a MacBook with either M1 or M2 chip you must check the following experimental feature in Docker Desktop: Settings -> Features in development -> Use Rosetta for x86/amd64 emulation on Apple Silicon
```

Thank you!

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java (hunk 9 above)
Comment (mdiwakar) [reply src/main/java/com/eliasnogueira/driver/BrowserFactory.java:121]: done

File: src/main/java/com/eliasnogueira/driver/BrowserFactory.java (hunk 11 above)
Comment (mdiwakar) [reply src/main/java/com/eliasnogueira/driver/BrowserFactory.java:170]: done

**Comment by mdiwakar** [comment]
done

**Comment by mdiwakar** [comment]
done

**Comment by mdiwakar** [comment]
@eliasnogueira , not sure why the PR is closed, Please merge the changes.
